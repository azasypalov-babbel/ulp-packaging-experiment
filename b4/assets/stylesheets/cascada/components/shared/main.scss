@use "sass:math";
// Layout

%responsive-container {
  margin: 0 auto;
  width: 90%;

  @include media(medium) {
    max-width: 940px;
  }
  @include media(large) {
    // fixed width for large and beyond
    max-width: none;
    width: 940px;
  }
  @include media(extra-large) {
    width: 1170px;
  }
}

// Typography

%trainer-page-text {
  font-size: math.div(16, 15) * 1em;
  margin-bottom: rem(24px);
  text-align: center;
}

%centered-columns {
  box-sizing: border-box;
  display: flex;
  justify-content: center;
  flex-direction: column;
}

// UI

%bordered-box {
  border-radius: $border-radius;
  background-color: $white;
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.16),
              0 0 2px 0 rgba(0, 0, 0, 0.12),
              inset 1px 1px 0 0 rgba(0, 0, 0, 0.08);
  color: $space-gray;
}

// Cascada override for Safari in iOS 10

.btn--primary,
.btn {
  display: inline-block;
}
