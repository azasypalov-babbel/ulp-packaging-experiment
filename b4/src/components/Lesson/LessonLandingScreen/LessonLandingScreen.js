import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import mediaUrlService from '../../../services/mediaUrlService';
import {
  StyledPageWrapper,
  StyledWrapper,
  StyledFlexWrapper,
  StyledHeroImage,
  StyledTextWrapper,
  StyledParagraph,
  StyledH1,
  StyledH3,
  StyledButton
} from './styles';
import withTranslations from '../../shared/withTranslations';

export const LessonLandingScreen = ({
  onConfirm,
  course,
  lesson,
  translations,
  track,
  learnLanguageAlpha3,
  locale
}) => {
  const imageUrl = mediaUrlService.imageURL(
    lesson.image.id,
    '1024x768',
    'jpg',
    'landscape'
  );

  useEffect(() => {
    /* eslint-disable camelcase */
    track({
      event: 'product:view_shown',
      version: 1,
      payload: {
        view_name: 'lesson_landing_page',
        origin: 'lesson_player',
        locale: locale,
        learn_language_alpha3: learnLanguageAlpha3
      }
    });
    /* eslint-enable camelcase */
  }, [track]);

  return (
    <StyledPageWrapper>
      <StyledHeroImage src={imageUrl} />
      <StyledWrapper>
        <StyledTextWrapper>
          <StyledH3 fontSize={'small'} color={'spaceGrayW28'}>
            {course.title}
          </StyledH3>
          <StyledH1
            fontSize={'large'}
            fontFamily={'fontLeituraBold'}
            color={'spaceGray'}
          >
            {lesson.title}
          </StyledH1>
          <StyledH3>{lesson.description}</StyledH3>
          <StyledFlexWrapper>
            <StyledParagraph
              fontSize={'small'}
              color={'spaceGray'}
              dangerouslySetInnerHTML={{
                __html: lesson.detailedDescriptionHtml
              }}
            />
            <StyledButton primary onClick={onConfirm}>
              {translations.button}
            </StyledButton>
          </StyledFlexWrapper>
        </StyledTextWrapper>
      </StyledWrapper>
    </StyledPageWrapper>
  );
};

LessonLandingScreen.propTypes = {
  course: PropTypes.shape({
    title: PropTypes.string.isRequired
  }).isRequired,
  lesson: PropTypes.shape({
    image: PropTypes.shape({
      id: PropTypes.string.isRequired
    }).isRequired,
    title: PropTypes.string.isRequired,
    detailedDescriptionHtml: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired
  }).isRequired,
  onConfirm: PropTypes.func.isRequired,
  translations: PropTypes.shape({
    button: PropTypes.string.isRequired
  }),
  track: PropTypes.func.isRequired,
  learnLanguageAlpha3: PropTypes.string.isRequired,
  locale: PropTypes.string.isRequired
};

const getTranslations = (translate) => ({
  button: translate('landing_screen.start_lesson')
});

export default withTranslations(getTranslations)(LessonLandingScreen);
