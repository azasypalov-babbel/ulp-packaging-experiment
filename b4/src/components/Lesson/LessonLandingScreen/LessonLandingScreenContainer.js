import { connect } from 'react-redux';
import { compose } from 'redux';
import { track } from '../../../dux/tracker/actions';
import * as contentSelectors from '../../../dux/content/selectors';

import LessonLandingScreen from './LessonLandingScreen';

const mapStateToProps = (state) => {
  const { content, lesson: { learningActivityId }, session } = state;
  const course = contentSelectors.currentCourse(content, learningActivityId);
  const lesson = contentSelectors.currentLesson(content, learningActivityId);

  return {
    course,
    lesson,
    learnLanguageAlpha3: session.learnLanguageAlpha3,
    locale: session.locale
  };
};

const mapDispatchToProps = {
  track
};

export default compose(
  connect(mapStateToProps, mapDispatchToProps)
)(LessonLandingScreen);
