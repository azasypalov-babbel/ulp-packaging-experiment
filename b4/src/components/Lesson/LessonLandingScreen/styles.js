import styled from 'styled-components';
import Text from '../../shared/Text';
import Button from '../../shared/Button/Button';

export const StyledPageWrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

export const StyledHeroImage = styled.div`
  width: 100%;
  max-width: 1024px;
  margin: auto;
  flex: 0 1 70%;
  background: center / cover no-repeat url(${(props) => props.src});
  z-index: 1;
`;

export const StyledWrapper = styled.div`
  flex: 1 0 30%;
  box-shadow: 0 0 0.15rem 0.08rem ${(props) => props.theme.cascada.silver};
`;

export const StyledTextWrapper = styled.div`
  max-width: 1024px;
  margin: auto;
  padding: 2.5rem 1.5rem;
  box-sizing: border-box;
`;

export const StyledFlexWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  flex-wrap: wrap;
`;

export const StyledH3 = styled(Text)`
  margin-bottom: 1.6rem;
`;

StyledH3.defaultProps = {
  as: 'h3'
};

export const StyledH1 = styled(Text)`
  margin-bottom: 0.4rem;
`;

StyledH3.defaultProps = {
  as: 'h1'
};

export const StyledParagraph = styled(Text)`
  line-height: 1.6;
`;

StyledParagraph.defaultProps = {
  as: 'p'
};

export const StyledButton = styled(Button)`
  margin-top: 1rem;
`;
