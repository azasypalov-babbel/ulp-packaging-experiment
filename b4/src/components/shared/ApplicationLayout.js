import styled, { css } from 'styled-components';
import { isiOS, isWebview } from '../../lib/features';

const fixNavigationBar = css`
  & .loy .loy-cascada-navbar {
    position: fixed;
    top: 0;
    padding-top: env(safe-area-inset-top);

    &:after {
      top: initial;
    }
  }
  & > main {
    margin-top: calc(${(props) => props.theme.size.navbar.xsmall.height} + env(safe-area-inset-top));
  }
`;

const disableSelection = css`
  user-select: none;
  -webkit-touch-callout: none;
`;

const ApplicationLayout = styled.div`
  display: flex;
  flex-direction: column;

  @media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
    height: 100vh; /* IE11 does not respect min-height for calculating flex */
  }

  min-height: 100vh;
  ${isiOS() && css`
    min-height: -webkit-fill-available; /* To reduce the page size if possible when software keyboard is open */
  `}

  ${isWebview() && fixNavigationBar};
  ${isWebview() && disableSelection};

  & > main {
    position: relative;
    flex: 1 0 auto; /* main should be least the height of it's content */

    display: flex;

    & > * {
      flex: 1 1 100%; /* make children 100% width, and shrink if screen width is too small */
    }
  }

  .hide-for-tiny {
    @media (max-width: 413px) {
      display: none;
    }
  }

  .fixed-for-small {
    @media (max-width: 767px) {
      position: fixed;
    }
  }

  .hide-for-small {
    @media (max-width: 767px) {
      display: none;
    }
  }
`;

export default ApplicationLayout;
