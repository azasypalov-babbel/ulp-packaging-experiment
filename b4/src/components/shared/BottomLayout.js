import React, { useEffect, useLayoutEffect, useState } from 'react';
import { getDisplayName } from '../getDisplayName';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { createPortal } from 'react-dom';
import useResizeObserver from './hooks/useResizeObserver';
import { getBottomInset, scrollIntoSafeArea, setBottomInset } from './scrollIntoSafeArea';
import { connect } from 'react-redux';

/**
 * This area is for any elements that are fixed to the bottom of the page
 * Give components `order` css property to organise them.
 */

const mountPointId = 'bottom-layout-mount-point';

export const StyledBottomLayout = styled.div`
  position: relative;

  @supports (position: sticky) {
    position: sticky;
    height: ${(props) => props.internalHeight}px;
  }

  bottom: 0;
  left: 0;
  right: 0;

  flex: 0 0 auto;
  z-index: 99;

  pointer-events: none;

  & > * {
    @supports (position: sticky) {
      position: absolute;
    }

    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    flex-direction: column;

    pointer-events: all;
  }
`;

export const bottomLayoutConfig = {
  enableScrollAdjustment: true,
  enableShrinking: false
};

const BottomLayoutComponent = ({ children, id }) => {
  const [internalHeight, setInternalHeight] = useState(0);

  const onResize = ({ height }) => {
    const isGrowing = height > internalHeight;

    if (isGrowing) {
      /**
       * InternalHeight is used to reserve space in the DOM.
       * It can only increase in height, to avoid making the page jump.
       */
      setInternalHeight(height);
    } else if (height !== 0 && bottomLayoutConfig.enableShrinking) {
      setInternalHeight(height);
    }

    const shouldAdjustScroll = height > getBottomInset();

    /**
     * Height of this element is used to calculate a safe area for scrolling
     * ie. so that content is not covered by anything in BottomLayout
     */
    setBottomInset(height);

    if (shouldAdjustScroll && bottomLayoutConfig.enableScrollAdjustment) {
      /**
       * Delay scroll until after next paint so that elements are in their final place
       */
      requestAnimationFrame(() => scrollIntoSafeArea());
    }
  };

  useEffect(() => {
    /**
     * Reset internal height when a the trainer changes or sequence starts or ends
     */
    setInternalHeight(0);
  }, [id]);

  const elementRef = useResizeObserver(onResize);

  return (
    <StyledBottomLayout internalHeight={internalHeight}>
      <div
        ref={elementRef}
        id={mountPointId}
      >
        {children}
      </div>
    </StyledBottomLayout>
  );
};

BottomLayoutComponent.propTypes = {
  children: PropTypes.node,
  id: PropTypes.string
};

const mapStateToProps = ({ sequence }) => ({
  id: [sequence.currentTrainerIndex, sequence.started, sequence.completed].join()
});

export const BottomLayout = connect(mapStateToProps)(BottomLayoutComponent);

export const renderInBottomLayout = (WrappedComponent) => {
  const RenderInBottomLayout = (props) => {
    const [element, setElement] = useState(document.getElementById(mountPointId));

    useLayoutEffect(() => {
      const domNode = document.getElementById(mountPointId);
      if (!domNode) {
        // eslint-disable-next-line max-len
        console.warn(`DOM element with id ${mountPointId} not found. This is a requirement for a bottom layout to be properly rendered.`);
      }
      setElement(domNode);
    }, []);

    if (element === null) {
      return null;
    }

    return createPortal(
      <WrappedComponent {...props} />,
      element
    );
  };

  RenderInBottomLayout.displayName = `RenderInBottomLayout(${getDisplayName(WrappedComponent)})`;

  return RenderInBottomLayout;
};
