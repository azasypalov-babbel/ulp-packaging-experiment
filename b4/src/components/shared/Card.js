import { transparentize, rem } from 'polished';
import styled, { css } from 'styled-components';

const Card = styled.div`
  box-sizing: border-box;
  overflow: hidden;

  background: ${(props) => props.theme.cascada.white};

  border-radius: ${rem('4px')};

  transition: box-shadow 0.3s ease;

  box-shadow: 0 ${rem('2px')} ${rem('20px')} 0 ${(props) => transparentize(0.9, props.theme.cascada.spaceGrayW15)};

  outline: none;

  ${(props) => Boolean(props.onClick) && css`

    cursor: pointer;

    &:hover, &:focus {
      box-shadow: 0 ${rem('8px')} ${rem('40px')} 0 ${(props) => transparentize(0.8, props.theme.cascada.spaceGrayW15)};
    }

    &:active {
      box-shadow: 0 ${rem('2px')} ${rem('10px')} 0 ${(props) => transparentize(0.7, props.theme.cascada.spaceGrayW15)};
    }
  `}
`;

export default Card;
