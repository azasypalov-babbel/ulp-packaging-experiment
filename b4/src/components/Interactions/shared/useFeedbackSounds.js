import { useCallback, useContext, useEffect, useReducer, useRef } from 'react';
import { ServiceContext } from '@src/components/shared/withServices';
import incorrectFeedbackSound from '@assets/sounds/babbel_wrong.mp3';
import correctFeedbackSound from '@assets/sounds/babbel_correct.mp3';
import { useSoundPlayer } from '@src/components/Trainer/shared/useSoundPlayer';
import { checks } from '@src/services/soundService';

const useFeedbackSounds = ({ skipOnActiveSound = true, resetDependencies = [] } = {}) => {
  const { settingsService, soundService } = useContext(ServiceContext);
  const correctSoundInstance = useSoundPlayer(correctFeedbackSound);
  const incorrectSoundInstance = useSoundPlayer(incorrectFeedbackSound);
  const [, forceUpdate] = useReducer((s) => !s, true);
  const skippedFeedbackRef = useRef(false);

  useEffect(() => {
    skippedFeedbackRef.current = false;
    correctSoundInstance.reset();
    incorrectSoundInstance.reset();
  }, [correctSoundInstance, incorrectSoundInstance, ...resetDependencies]);

  const hasPlayed = checks.hasPlayed(correctSoundInstance.getState())
    || checks.hasPlayed(incorrectSoundInstance.getState());

  return [
    useCallback((solved) => {
      const isPlayingAnySound = soundService.getPlayingInstances().length > 0;
      if ((skipOnActiveSound && isPlayingAnySound) || !settingsService.isFeedbackSoundEnabled) {
        skippedFeedbackRef.current = true;
        forceUpdate();
        return;
      }
      const instance = solved ? correctSoundInstance : incorrectSoundInstance;
      instance.play();
    }, [skipOnActiveSound, soundService, settingsService, correctSoundInstance, incorrectSoundInstance]),
    hasPlayed || skippedFeedbackRef.current
  ];
};

export default useFeedbackSounds;
