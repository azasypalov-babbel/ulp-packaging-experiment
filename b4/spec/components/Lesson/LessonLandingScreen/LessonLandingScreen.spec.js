import React from 'react';
import { shallow } from 'enzyme';
import { LessonLandingScreen } from '../../../../src/components/Lesson/LessonLandingScreen/LessonLandingScreen';
import { mountWithTheme } from '../../shared/themeMock';

describe('<LessonLandingScreen />', () => {
  const defaultProps = {
    locale: 'en',
    learnLanguageAlpha3: 'FRA',
    onConfirm: jest.fn(),
    course: {
      title: 'mock course title'
    },
    lesson: {
      title: 'mock lesson title',
      detailedDescriptionHtml: '<div>mock lesson description hmtl</div>',
      description: 'mock lesson description',
      image: {
        id: 'mocklessonimageid'
      }
    },
    translations: {
      button: 'Start lesson'
    },
    track: jest.fn()
  };

  describe('Rendering', () => {
    test('it renders correctly', () => {
      const component = shallow(
        <LessonLandingScreen {...defaultProps} />
      );

      expect(component).toMatchSnapshot();
    });
  });

  describe('Integration', () => {
    test('it calls `onConfirm` when `button` is clicked', () => {
      const component = mountWithTheme(
        <LessonLandingScreen {...defaultProps} />
      );

      component.find('button').simulate('click');

      expect(defaultProps.onConfirm).toHaveBeenCalledTimes(1);
    });
  });
});
