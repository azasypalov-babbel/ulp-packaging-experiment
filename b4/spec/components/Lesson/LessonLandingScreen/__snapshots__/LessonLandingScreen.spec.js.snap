// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LessonLandingScreen /> Rendering it renders correctly 1`] = `
<styled.div>
  <styled.div
    src="https://images.babbel.com/v1.0.0/images/mocklessonimageid/variations/landscape/resolutions/1024x768.jpg"
  />
  <styled.div>
    <styled.div>
      <Styled(styled.span)
        as="h1"
        color="spaceGrayW28"
        fontSize="small"
      >
        mock course title
      </Styled(styled.span)>
      <Styled(styled.span)
        color="spaceGray"
        fontFamily="fontLeituraBold"
        fontSize="large"
      >
        mock lesson title
      </Styled(styled.span)>
      <Styled(styled.span)
        as="h1"
      >
        mock lesson description
      </Styled(styled.span)>
      <styled.div>
        <Styled(styled.span)
          as="p"
          color="spaceGray"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<div>mock lesson description hmtl</div>",
            }
          }
          fontSize="small"
        />
        <Styled(WithKeypress(styled.button))
          onClick={[MockFunction]}
          primary={true}
        >
          Start lesson
        </Styled(WithKeypress(styled.button))>
      </styled.div>
    </styled.div>
  </styled.div>
</styled.div>
`;
